### PR TITLE
VR/AR buttons to ES6 classes

### PR DESCRIPTION
**Pull request by ivnnv**
My very first addition!
Tested in Chrome in some of the examples on both with and without the webXR mozilla extension (so I could test the "not available" state.

`onSessionStarted`, `onSessionEnded`, and the eventListeners need to be arrow functions to be able to keep the `this` reference and avoid having to  `.bind()`

### CODE REVIEW

File: examples/jsm/webxr/ARButton.js [hunk 1]
```diff
@@ -1,161 +1,195 @@
...
+		this.message = null;
 
-			var currentSession = null;
+		this.currentSession = null;
 
-			function onSessionStarted( session ) {
+	}
 
-				session.addEventListener( 'end', onSessionEnded );
+	createButton( renderer, sessionInit = {} ) {
```
Comment (Mugen87) [inline examples/jsm/webxr/ARButton.js:14]: I think it's better to just make `createButton()` to a static function. This code:
```js
const newARButton = new ARButton();
```
should not be necessary.

Comment (ivnnv) [reply examples/jsm/webxr/ARButton.js:14]: I think I don't get it and I just tried your suggestion.
If we define createButton as a static method we lose the `this` reference and would be needed to `new VRButton` when importing the button and that will break any existing app which expects to get a full (somehow initialized in the constructor) instance and then calling its .createButton which has references to any this.func needed, even the listeners:
This change is fully backwards compatible.

Comment (Mugen87) [reply examples/jsm/webxr/ARButton.js:14]: An app does not call `new VRButton`.  The intended usage is:
https://github.com/mrdoob/three.js/blob/29ec65b5223eb4a0cd155a6085e6ac7999b748d8/examples/webxr_vr_ballshooter.html#L90

This code seems to work just fine:
```js
class VRButton {

	static createButton( renderer, options ) {

		if ( options ) {

			console.error( 'THREE.VRButton: The "options" parameter has been removed. Please set the reference space type via renderer.xr.setReferenceSpaceType() instead.' );

		}

		function showEnterVR( /*device*/ ) {

			var currentSession = null;

			function onSessionStarted( session ) {

				session.addEventListener( 'end', onSessionEnded );

				renderer.xr.setSession( session );
				button.textContent = 'EXIT VR';

				currentSession = session;

			}

			function onSessionEnded( /*event*/ ) {

				currentSession.removeEventListener( 'end', onSessionEnded );

				button.textContent = 'ENTER VR';

				currentSession = null;

			}

			//

			button.style.display = '';

			button.style.cursor = 'pointer';
			button.style.left = 'calc(50% - 50px)';
			button.style.width = '100px';

			button.textContent = 'ENTER VR';

			button.onmouseenter = function () {

				button.style.opacity = '1.0';

			};

			button.onmouseleave = function () {

				button.style.opacity = '0.5';

			};

			button.onclick = function () {

				if ( currentSession === null ) {

					// WebXR's requestReferenceSpace only works if the corresponding feature
					// was requested at session creation time. For simplicity, just ask for
					// the interesting ones as optional features, but be aware that the
					// requestReferenceSpace call will fail if it turns out to be unavailable.
					// ('local' is always available for immersive sessions and doesn't need to
					// be requested separately.)

					var sessionInit = { optionalFeatures: [ 'local-floor', 'bounded-floor', 'hand-tracking' ] };
					navigator.xr.requestSession( 'immersive-vr', sessionInit ).then( onSessionStarted );

				} else {

					currentSession.end();

				}

			};

		}

		function disableButton() {

			button.style.display = '';

			button.style.cursor = 'auto';
			button.style.left = 'calc(50% - 75px)';
			button.style.width = '150px';

			button.onmouseenter = null;
			button.onmouseleave = null;

			button.onclick = null;

		}

		function showWebXRNotFound() {

			disableButton();

			button.textContent = 'VR NOT SUPPORTED';

		}

		function stylizeElement( element ) {

			element.style.position = 'absolute';
			element.style.bottom = '20px';
			element.style.padding = '12px 6px';
			element.style.border = '1px solid #fff';
			element.style.borderRadius = '4px';
			element.style.background = 'rgba(0,0,0,0.1)';
			element.style.color = '#fff';
			element.style.font = 'normal 13px sans-serif';
			element.style.textAlign = 'center';
			element.style.opacity = '0.5';
			element.style.outline = 'none';
			element.style.zIndex = '999';

		}

		if ( 'xr' in navigator ) {

			var button = document.createElement( 'button' );
			button.id = 'VRButton';
			button.style.display = 'none';

			stylizeElement( button );

			navigator.xr.isSessionSupported( 'immersive-vr' ).then( function ( supported ) {

				supported ? showEnterVR() : showWebXRNotFound();

			} );

			return button;

		} else {

			var message = document.createElement( 'a' );

			if ( window.isSecureContext === false ) {

				message.href = document.location.href.replace( /^http:/, 'https:' );
				message.innerHTML = 'WEBXR NEEDS HTTPS'; // TODO Improve message

			} else {

				message.href = 'https://immersiveweb.dev/';
				message.innerHTML = 'WEBXR NOT AVAILABLE';

			}

			message.style.left = 'calc(50% - 90px)';
			message.style.width = '180px';
			message.style.textDecoration = 'none';

			stylizeElement( message );

			return message;

		}

	}

}

export { VRButton };

```

Comment (ivnnv) [reply examples/jsm/webxr/ARButton.js:14]: Ahh I get what you mean, putting everything inside createButton, but thats not very ES6ish tbh...
What I wanted to create is a modern class as if we were creating it from scratch, while keeping it backward compatible as  having the fact that we need to live forever with the .createButton initialization approach.

Comment (Mugen87) [reply examples/jsm/webxr/ARButton.js:14]: > but thats not very ES6ish tbh

Well, that's a matter of taste. Not everything has to be written as a class. 

TBH, I would keep the code as it is since I personally see no improvement in your approach. It's even worse than before since now a lot of methods are public. `createButton()` should be the only public interface that the button modules provide.

Comment (ivnnv) [reply examples/jsm/webxr/ARButton.js:14]: > Not everything has to be written as a class

The fact this PR exists is because I thought you were just in the process of modernizing the codebase and transforming every component to a class, this approach is more ES6ish than just wrapping the same old fashioned function inside a dumb class container.

> Declaring non-public related stuff with an _ prefix is an acceptable tradeoff.

So this PR would be accepted defining non-createButton functions prefixed with underscore?

Comment (Mugen87) [reply examples/jsm/webxr/ARButton.js:14]: Originally the class migration was started to improve tree shaking capabilities. But it turned out, this can be solved differently. I'm not sure anymore we want to go down the path and convert the entire codebase to classes if tree shaking can be improved differently.

> same old fashioned function 

Your argumentation seems a bit biased to me. Clean and modern code does not need to reside in a class. 

> So this PR would be accepted defining non-createButton functions prefixed with underscore?

That depends on @mrdoob preferences. I personally see no need for a change.

Comment (Mugen87) [reply examples/jsm/webxr/ARButton.js:14]: @mrdoob In `WebGPURenderer` and its new components, I've decided to start with the class approach. However, if you prefer the existing style of `WebGLRenderer`, the code can be refactored. I know that not every dev is a fan of pseudo-private properties and methods achieved via the `_` prefix.

Comment (mrdoob) [reply examples/jsm/webxr/ARButton.js:14]: > The fact this PR exists is because I thought you were just in the process of modernizing the codebase and transforming every component to a class

Yes, but we stopped when we realised we would be poluting the API with `_` variables because browsers don't support private properties in classes yet.

> @mrdoob In `WebGPURenderer` and its new components, I've decided to start with the class approach. However, if you prefer the existing style of `WebGLRenderer`, the code can be refactored.

Using `_` in `WebGPURenderer` is fine. Eventually we'll need to be converted them into `this.#`. Hopefully browsers that support webgpu will also support private properties in classes.

Comment (ivnnv) [reply examples/jsm/webxr/ARButton.js:14]: Ok guys I get your point, I have modified the VRButton to follow the static way, if you like it i'll modify the ARButton too, otherwise feel free to just close the PR.

Anyway thanks for the amazing work on the library.

Comment (mrdoob) [reply examples/jsm/webxr/ARButton.js:14]: `VRButton` is looking better. I think `() =>` are no longer needed though?

File: examples/jsm/webxr/ARButton.js [hunk 2]
```diff
@@ -1,40 +1,38 @@
...
+			currentSession = session;
 
-				currentSession = null;
+		}
 
-			}
+		function showStartAR( /*device*/ ) {
 
-			//
+			this.currentSession = null;
```
Comment (mrdoob) [inline examples/jsm/webxr/ARButton.js:35]: You have a `let currentSession` and `this.currentSession`.

File: examples/jsm/webxr/ARButton.js [hunk 3]
```diff
@@ -97,40 +96,54 @@ var ARButton = {
...
-			element.style.border = '1px solid #fff';
-			element.style.borderRadius = '4px';
-			element.style.background = 'rgba(0,0,0,0.1)';
-			element.style.color = '#fff';
-			element.style.font = 'normal 13px sans-serif';
-			element.style.textAlign = 'center';
-			element.style.opacity = '0.5';
-			element.style.outline = 'none';
-			element.style.zIndex = '999';
+			const addedStyles = {
```
Comment (mrdoob) [inline examples/jsm/webxr/ARButton.js:99]: Why did you change this?

File: examples/jsm/webxr/ARButton.js [hunk 4]
```diff
@@ -97,40 +96,54 @@ var ARButton = {
...
 			button.id = 'ARButton';
 			button.style.display = 'none';
 
-			stylizeElement( button );
-
-			navigator.xr.isSessionSupported( 'immersive-ar' ).then( function ( supported ) {
+			navigator.xr.isSessionSupported( 'immersive-ar' ).then( supported => {
 
-				supported ? showStartAR() : showARNotSupported();
+				if ( supported )
```
Comment (mrdoob) [inline examples/jsm/webxr/ARButton.js:131]: Why did you change this?

Comment (ivnnv) [reply examples/jsm/webxr/ARButton.js:131]: eslint (and personal) preference, dont you think its more readable?

File: examples/jsm/webxr/ARButton.js (hunk 3 above)
Comment (ivnnv) [reply examples/jsm/webxr/ARButton.js:99]: Eres rapido! :P
I think its more encapsulated than straight manipulation of the original object, define the to-be-added properties with a nice name that the reader knows the purpose in advance and then iterate thru it, ready for a (if) a bigger refactor.

**Comment by mrdoob** [comment]
It seems like you're not just changing the code to use ES6 classes, but you're making stylistic decisions on other parts of the code in the same PR making the review harder than it should.

File: examples/jsm/webxr/ARButton.js (hunk 2 above)
Comment (ivnnv) [reply examples/jsm/webxr/ARButton.js:35]: good catch, will change

File: examples/jsm/webxr/ARButton.js (hunk 4 above)
Comment (mrdoob) [reply examples/jsm/webxr/ARButton.js:131]: I just don't think this change should be part of this PR.

File: examples/jsm/webxr/ARButton.js (hunk 3 above)
Comment (mrdoob) [reply examples/jsm/webxr/ARButton.js:99]: I don't think this change should be part of this PR either.

**Comment by mrdoob** [comment]
Have you tested this code?

File: examples/jsm/webxr/ARButton.js (hunk 3 above)
Comment (ivnnv) [reply examples/jsm/webxr/ARButton.js]: got it, will PR any additional changes on a different one and maybe an extra feature

File: examples/jsm/webxr/ARButton.js [hunk 5]
```diff
@@ -1,40 +1,38 @@
...
 
-		function showStartAR( /*device*/ ) {
+		let currentSession = null;
 
-			var currentSession = null;
+		const button = document.createElement( 'button' );
+		const message = document.createElement( 'a' );
 
-			function onSessionStarted( session ) {
+		function onSessionEnded( /*event*/ ) {
```
Comment (mrdoob) [inline examples/jsm/webxr/ARButton.js:10]: More questions... Why did you change the order of the functions?

Comment (ivnnv) [reply examples/jsm/webxr/ARButton.js:10]: eslint complains and suggest to declare functions before they are used, so they are all now declared before they have been called

**Comment by mrdoob** [comment]
Something went wrong: https://github.com/mrdoob/three.js/pull/20302/files

Sorry, I'm going to close this and do the change myself.

Hopefully next PR goes more smoothly!

**Comment by ivnnv** [comment]
Yep, I tried to clean the git history mergin against dev and the PR got dirty, nevermind, maybe next time

**Comment by mrdoob** [comment]
https://github.com/mrdoob/three.js/commit/853f44510b3bdeb88946ae954c5981e39dbee5af
